File: backend/app/schemas/copilot.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from app.models.enums import (
    CopilotRole,
    CopilotPlanStatus,
    CopilotToolStatus,
    DataStatusType,
)


class ToolDefinitionSchema(BaseModel):
    name: str
    description: str
    input_schema: Dict[str, Any] = Field(default_factory=dict)
    output_schema: Dict[str, Any] = Field(default_factory=dict)
    permission: str = "READ" # "READ", "ANALYZE", "WRITE"
    timeout_sec: int = 10
    data_requirements: List[str] = Field(default_factory=list)
    source: str = "FREIGHT_IQ_SERVICE"


class ToolResultSchema(BaseModel):
    tool_name: str
    status: CopilotToolStatus
    data: Optional[Dict[str, Any]] = None
    source: str = "FREIGHT_IQ"
    data_status: DataStatusType = DataStatusType.SYNTHETIC
    dataset_version_id: Optional[str] = None
    model_version: Optional[str] = None
    observed_at: Optional[str] = None
    last_updated: Optional[str] = None
    currency: str = "USD"
    unit: Optional[str] = None
    assumption: Optional[str] = None
    execution_time_ms: float = 0.0
    error: Optional[str] = None


class CopilotPlanStepSchema(BaseModel):
    step_number: int
    tool_name: str
    reason: str
    arguments: Dict[str, Any] = Field(default_factory=dict)
    status: CopilotToolStatus = CopilotToolStatus.PENDING


class CopilotPlanSchema(BaseModel):
    id: str
    session_id: str
    user_goal: str
    steps: List[CopilotPlanStepSchema] = Field(default_factory=list)
    status: CopilotPlanStatus = CopilotPlanStatus.PLANNED
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CopilotPlanRequest(BaseModel):
    query: str
    session_id: Optional[str] = None
    cargo_request_id: Optional[str] = None
    vessel_id: Optional[str] = None
    voyage_id: Optional[str] = None


class CopilotToolCallSchema(BaseModel):
    id: str
    session_id: str
    tool_name: str
    arguments: Dict[str, Any] = Field(default_factory=dict)
    status: CopilotToolStatus
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    execution_time_ms: Optional[float] = None
    error: Optional[str] = None
    result: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class StructuredCopilotResponse(BaseModel):
    """
    Standardized traceable JSON response structure synthesized by Copilot
    from verified tool results.
    """
    summary: str
    findings: List[str] = Field(default_factory=list)
    decision_context: Dict[str, Any] = Field(default_factory=dict)
    evidence: List[Dict[str, Any]] = Field(default_factory=list)
    risks: List[Dict[str, Any]] = Field(default_factory=list)
    economics: Dict[str, Any] = Field(default_factory=dict)
    assumptions: List[str] = Field(default_factory=list)
    uncertainties: List[str] = Field(default_factory=list)
    data_quality: Dict[str, Any] = Field(default_factory=dict)
    actions: List[Dict[str, Any]] = Field(default_factory=list)


class CopilotMessageSchema(BaseModel):
    id: str
    session_id: str
    role: CopilotRole
    content: str
    structured_response: Optional[StructuredCopilotResponse] = None
    plan_id: Optional[str] = None
    created_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class CopilotSessionSchema(BaseModel):
    id: str
    title: str
    cargo_request_id: Optional[str] = None
    vessel_id: Optional[str] = None
    voyage_id: Optional[str] = None
    context_data: Optional[Dict[str, Any]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    message_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class CopilotSessionDetailSchema(BaseModel):
    session: CopilotSessionSchema
    messages: List[CopilotMessageSchema] = Field(default_factory=list)
    latest_plan: Optional[CopilotPlanSchema] = None
    tool_calls: List[CopilotToolCallSchema] = Field(default_factory=list)


class CopilotChatRequest(BaseModel):
    message: str
    session_id: Optional[str] = None
    stream: bool = False
    cargo_request_id: Optional[str] = None
    vessel_id: Optional[str] = None
    voyage_id: Optional[str] = None
    scenario_preset: Optional[str] = None

File: src/lib/api/copilot.ts
import { apiClient } from './client';

export type CopilotRole = 'user' | 'assistant' | 'system';
export type CopilotPlanStatus = 'PLANNED' | 'EXECUTING' | 'COMPLETED' | 'PARTIAL' | 'FAILED';
export type CopilotToolStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';

export interface StructuredCopilotResponse {
  summary: string;
  findings: string[];
  decision_context: Record<string, any>;
  evidence: Array<{
    dimension: string;
    source: string;
    status: string;
    metric: string;
  }>;
  risks: Array<{
    type: string;
    severity: string;
    active_count?: number;
    cost_usd?: number;
    description: string;
  }>;
  economics: {
    total_voyage_cost_usd: number;
    cost_per_mt: number;
    freight_cost_usd?: number;
    bunker_cost_usd?: number;
    port_dues_usd?: number;
    time_charter_usd?: number;
    delay_exposure_usd?: number;
    repositioning_usd?: number;
  };
  assumptions: string[];
  uncertainties: string[];
  data_quality: {
    overall_status: string;
    confidence_score: number;
    provenance_verified: boolean;
  };
  actions: Array<{
    label: string;
    route: string;
  }>;
}

export interface CopilotMessage {
  id: string;
  session_id: string;
  role: CopilotRole;
  content: string;
  structured_response?: StructuredCopilotResponse;
  plan_id?: string;
  created_at?: string;
}

export interface CopilotSession {
  id: string;
  title: string;
  cargo_request_id?: string;
  vessel_id?: string;
  voyage_id?: string;
  context_data?: Record<string, any>;
  created_at?: string;
  updated_at?: string;
  message_count?: number;
  cargo_requirement?: {
    title: string;
    quantity_mt: number;
    cargo_type: string;
  };
  vessel?: {
    name: string;
    vessel_class: string;
  };
}

export interface CopilotPlanStep {
  step_number: number;
  tool_name: string;
  reason: string;
  arguments: Record<string, any>;
  status: CopilotToolStatus;
}

export interface CopilotPlan {
  id: string;
  session_id: string;
  user_goal: string;
  status: CopilotPlanStatus;
  steps: CopilotPlanStep[];
  created_at?: string;
  completed_at?: string;
}

export interface ToolTraceItem {
  id: string;
  tool_name: string;
  arguments: Record<string, any>;
  status: CopilotToolStatus;
  execution_time_ms: number;
  source: string;
  data_status: string;
  result?: any;
  error?: string;
  started_at?: string;
  completed_at?: string;
}

export interface SessionTrace {
  session_id: string;
  title: string;
  total_tools_executed: number;
  total_execution_time_ms: number;
  latest_plan?: CopilotPlan;
  tools_trace: Array<{
    tool_name: string;
    status: string;
    execution_time_ms: number;
    source: string;
    data_status: string;
    error?: string;
  }>;
}

export interface ChatRequest {
  message: string;
  session_id?: string;
  stream?: boolean;
  cargo_request_id?: string;
  vessel_id?: string;
  voyage_id?: string;
  scenario_preset?: string;
}

export const copilotApi = {
  getSessions: async (limit = 50): Promise<CopilotSession[]> => {
    return apiClient<CopilotSession[]>(`/copilot/sessions?limit=${limit}`);
  },

  getSession: async (id: string): Promise<CopilotSession> => {
    return apiClient<CopilotSession>(`/copilot/sessions/${id}`);
  },

  getMessages: async (sessionId: string): Promise<CopilotMessage[]> => {
    return apiClient<CopilotMessage[]>(`/copilot/sessions/${sessionId}/messages`);
  },

  getTools: async (sessionId: string): Promise<ToolTraceItem[]> => {
    return apiClient<ToolTraceItem[]>(`/copilot/sessions/${sessionId}/tools`);
  },

  getTrace: async (sessionId: string): Promise<SessionTrace> => {
    return apiClient<SessionTrace>(`/copilot/sessions/${sessionId}/trace`);
  },

  previewPlan: async (query: string, sessionId?: string): Promise<{ query: string; steps: CopilotPlanStep[]; total_tools: number }> => {
    return apiClient(`/copilot/plan`, {
      method: 'POST',
      body: JSON.stringify({ query, session_id: sessionId }),
    });
  },

  chatSync: async (req: ChatRequest): Promise<any> => {
    return apiClient(`/copilot/chat`, {
      method: 'POST',
      body: JSON.stringify({ ...req, stream: false }),
    });
  },

  chatStream: async (
    req: ChatRequest,
    onStep: (step: any) => void,
    onMessage: (data: any) => void,
    onError: (err: any) => void,
    onDone: () => void
  ) => {
    const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://127.0.0.1:8000/api/v1';
    try {
      const response = await fetch(`${API_BASE_URL}/copilot/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'text/event-stream',
        },
        body: JSON.stringify({ ...req, stream: true }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error ${response.status}`);
      }

      if (!response.body) {
        throw new Error('ReadableStream not supported.');
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder('utf-8');
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n\n');
        buffer = lines.pop() || '';

        for (const block of lines) {
          if (!block.trim()) continue;
          const matchEvent = block.match(/event:\s*([^\n]+)/);
          const matchData = block.match(/data:\s*([^\n]+)/);

          const eventType = matchEvent ? matchEvent[1].trim() : 'message';
          const rawData = matchData ? matchData[1].trim() : '';

          if (!rawData) continue;

          try {
            const parsed = JSON.parse(rawData);
            if (eventType === 'step') {
              onStep(parsed);
            } else if (eventType === 'message') {
              onMessage(parsed);
            } else if (eventType === 'error') {
              onError(parsed);
            } else if (eventType === 'done') {
              onDone();
            }
          } catch (e) {
            console.error('Failed to parse SSE payload:', rawData, e);
          }
        }
      }
      onDone();
    } catch (err) {
      onError(err);
    }
  },
};
